src/unit.c

```diff
@@ -23,148 +23,63 @@
 #include <float.h>
 #include <string.h>
 
-int isUnitArithmetic(SEXP ua) {
-    return inherits(ua, "unit.arithmetic");
-}
-
-int isUnitList(SEXP ul) {
-    return inherits(ul, "unit.list");
-}
-
 /* Function to build a single-value unit SEXP internally.
  * Cannot build units requiring data as yet.
  */
 SEXP unit(double value, int unit) 
 {
-    SEXP u, units, classname;
-    PROTECT(u = ScalarReal(value));
-    PROTECT(units = ScalarInteger(unit));
-    /* NOTE that we do not set the "unit" attribute */
-    setAttrib(u, install("valid.unit"), units);
-    setAttrib(u, install("data"), R_NilValue);
-    PROTECT(classname = mkString("unit"));
-    classgets(u, classname);
-    UNPROTECT(3);
-    return u;
+  SEXP units = PROTECT(allocVector(VECSXP, 1));
+  SEXP u = SET_VECTOR_ELT(units, 0, allocVector(VECSXP, 3));
+  SET_VECTOR_ELT(u, 0, ScalarReal(value));
+  SET_VECTOR_ELT(u, 1, R_NilValue);
+  SET_VECTOR_ELT(u, 2, ScalarInteger(unit));
+  SEXP cl = PROTECT(mkString("unit"));
+  classgets(units, cl);
+  UNPROTECT(2);
+  return units;
+}
+
+int isSimpleUnit(SEXP unit) {
+	return inherits(unit, "simpleUnit");
 }
 
 /* Accessor functions for unit objects
  */
-
-/* 
- * This is an attempt to extract a single numeric value from
- * a unit.  This is ONLY designed for use on "simple" units
- * (i.e., NOT unitLists or unitArithmetics)
+/*
+ * This extracts the underlying scalar unit list structure from the unit vector
  */
+SEXP unitScalar(SEXP unit, int index) {
+	int i = index % LENGTH(unit);
+	if (isSimpleUnit(unit)) {
+		SEXP newUnit = PROTECT(allocVector(VECSXP, 3));
+		SET_VECTOR_ELT(newUnit, 0, Rf_ScalarReal(REAL(unit)[i]));
+		SET_VECTOR_ELT(newUnit, 1, R_NilValue);
+		SET_VECTOR_ELT(newUnit, 2, Rf_ScalarInteger(INTEGER(getAttrib(unit, install("unit")))[0]));
+		UNPROTECT(1);
+		return newUnit;
+	}
+	return VECTOR_ELT(unit, i);
```

Comment:
Maybe I am getting tired, but I can't get that line to make sense - doesn't 'unit' always have exactly 3 elements ?  Isn't this indexing on the wrong thing ?  Shouldn't this be the equivalent of lapply(unit, "[", i) ?

Reply:
In fact, most of the calls to unitScalar() I am seeing so far appear to do something else for "simpleUnit"s and only call unitScalar() for standard units (e.g., line 65 and line 70 below).  I am getting a bit confused (and tired) - any clarification gratefully received :)

Reply:
The intend is to get the underlying *bare* element of a unit vector, i.e. a 3 element list. If the object is a `simpleUnit` this must be constructed as it doesn't exist yet (`simpleUnit` is a numeric vector). If it is a regular `unit`, we simply return the element. It should be equal to `unit[[I]]`, not `unit[I]`, as we shave off the outer list...

It is true that in most cases we try to be smarter about `simpleUnit` objects, but I want this accessor to behave correctly no matter what we give it to protect us in the future. Usually we would extract the required information using `unitValue`, `unitUnit`, and `unitData`, but e.g. when summing we need the scalar to put into the data object of the new unit...

Reply:
I think I have this clearer now.  Having the new merged patch allowed me to build and see these new "unit" objects in the flesh, which helped a lot :)  Thanks!